README
Apache::MP3 Version 4.0

This module takes a hierarchy of directories containing MP3, MP4, Ogg
Vorbis, or Wav audio files and presents it as a browsable and
searchable song library for streaming over the web.  It requires the
Apache (version 2) web server, the mod_perl (version 2) embedded Perl
interpreter, Locale::Maketext and CGI::Session. In addition, you must
have any or all of the MP3::Info, MP4::Info,
Ogg::Vorbis/Ogg::Vorbis::Header::PurePerl and Audio::Wav modules.

Audio files are displayed in a list that shows the title, artist, duration
and bitrate.  Subdirectories are displayed with "CD" icons. The user can
download an MP3 file to disk by clicking on its title, stream it to an MP3
decoder by clicking on the "play" link. Users can also stream the entire
contents of a directory, or select a subset of songs to play.

A DEMO is available at http://www.modperl.com/Songs.  In this demo, stream
time is limited to 30 seconds in order to avoid copyright infringement.

See the "screenshots" directory for some pictures of Apache::MP3 in action.

See below for CHANGES.

INSTALLATION

1. Prequisites

This module requires Apache2's mod_perl, Locale::Maketext and
CGI::Session all of which are available on CPAN.  Optional are
MP3::Info, MP4::Info, Ogg::Vorbis/Ogg::Vorbis::Header::PurePerl and
Audio::Wav.

2. Configure MIME types

Apache must be configured to recognize the mp3 and MP3 extensions as MIME
type audio/mpeg.  Add the following to httpd.conf or srm.conf:

 AddType audio/mpeg mp3 MP3
 AddType audio/playlist m3u M3U

If you plan to use Shoutcast-style playlists, add the following as well:

 AddType audio/x-scpls pls PLS

If you plan to use Ogg Vorbis, add the following as well:

 AddType application/x-ogg ogg OGG

3. Install icons and stylesheet

This module uses a set of icons and a cascading stylesheet to generate its
song listings.  By default, the module expects to find them at the url
/apache_mp3.  Create a directory named apache_mp3 in your document root,
and copy into it the contents of the F<apache_mp3> directory from the
Apache::MP3 distribution.

You may change the location of this directory by setting the I<BaseDir>
configuration variable.  See the I<Customizing> section for more details.

4. Set Apache::MP3 to be the handler for the songs directory

In httpd.conf or access.conf, create a E<lt>LocationE<gt> or
E<lt>DirectoryE<gt> section, and make Apache::MP3 the handler for this
directory.  This example assumes you are using the URL /Songs as the
directory where you will be storing song files:

  <Location /Songs>
    SetHandler perl-script
    PerlHandler Apache::MP3
  </Location>

If you would prefer the song file listing that allows the user to sort it
in various ways, set the handler to use the Apache::MP3::Sorted subclass
instead.  A further elaboration is Apache::MP3::Playlist, which uses
cookies to manage a persistent playlist for the user.

5. Load MP3::Info in the Perl Startup file (optional)

For the purposes of faster startup and memory efficiency, you may load the
MP3::Info module at server startup time.  If you have a mod_perl "startup"
file, enter these lines:

  use MP3::Info;
  use Apache::MP3;

6. Set up song directory

Create a directory in the web server document tree that will contain the
audio files to be served.  The module recognizes and handles
subdirectories appropriately.  I suggest organizing directories
hierarchically by artist and/or album name.

If you place a file named "cover.jpg" in any of the directories, that
image will be displayed at the top of the directory listing.  You can use
this to display cover art.

If you place a list of .mp3 file names in a file with the .m3u extension,
it will be treated as a playlist and displayed to the user with a
distinctive icon.  Selecting the playlist icon will download the playlist
and stream its contents.  The playlist must contain relative file names,
but may refer to subdirectories, as in this example:

  # file: folk_favorites.m3u
  Never_a_Moment_s_Thought_v2.mp3
  Peter Paul & Mary - Leaving On A Jet Plane.mp3
  Simon and Garfunkel/Simon And Garfunkel - April Come She Will.mp3

7. Set up an information cache directory (optional)

In order to generate its MP3 listing, Apache::MP3 must open each sound
file, extract its header information, and close it.  This is time
consuming, particularly when recursively generating playlists across
multiple directories.  To speed up this process, Apache::MP3 has the
ability cache MP3 file information in a separate directory area.

To configure this, choose a directory that the Web server has write access
for, such as /usr/tmp.  Then add a configuration variable like the
following to the <Location> directive:

 PerlSetVar  CacheDir       /usr/tmp/mp3_cache

If the designated directory does not exist, Apache::MP3 will attempt to
create it, limited of course by the Web server's privileges.  You may need
to create the mp3_cache directory yourself if /usr/tmp is not world
writable.

TESTING IT

Open up the MP3 URL in your favorite browser.  You should be able to see
directory listings, and download and stream your songs.  If things don't
seem to be working, checking the server error log for messages.

Run "perldoc Apache::MP3" for more configuration information.

CHANGES

See Changes.

AUTHORS

Lincoln Stein, <lstein@cshl.org>


Changes
4.01
	-Playlists sorted alphabetically.
4.00
	-Now works *only* with Apache2/mod_perl2.
	-Removed searching functionality -- was not stable.
        -Removed MP3::Icecast dependency -- this module cannot produce headers
		needed by Apache2.

3.05	Mon Oct  6 10:09:47 EDT 2003
	-Put version back into Apache::Resample
	-Performance and functionality improvements to search function
	from Evan Harris.	

3.04 Fri Nov 29 16:33:29 EST 2002
	-Fixed incorrect HTML content type in the META tag.
	-Fixed incorrect playlist port numbers when running in non-virtual
		host, but using non-80 port.
	-Added javascript functions for use with help pages.
	-Chinese support from Autrijus Tang

3.03 Sun Oct 13 23:21:00 EDT 2002
	- Added generic filtering code to Apache::Resample contributed by
	Caleb Epstein.
	- Fixed bug in the playlist handling code that prevented individual files
	from streaming when [stream] link pressed.

3.01 Sun Aug 18 13:32:59 EDT 2002
	-Changed the way that $VERSION is defined in the localization modules
	in order to live peacefully with CPAN.

3.00 Fri Aug 16 00:12:33 EDT 2002
	- Extensive contribution from Sean Burke to provide localization and
		internationalization.

2.26 Wed Apr 17 11:03:24 EDT 2002
	- Added patch from Tom Hughes to correctly escape paths in the playlist.
	- Added patch from Tom Hughes to fetch the average bitrate for VBR Ogg
		files rather than the nominal bitrate (which is meaningless).

2.25 Tue Apr  9 12:35:44 EDT 2002
	- Added patch from Clemens Schrimpe to handle diacritical marks in
		CD icon names.
	- Added patch from John Regehr to get CDDB information from an 
	external index file, if present.
        - Put warning in the error_log if PerlSetupEnv has been set to off
        (thanks to Stas Bekman and Eric Hammond)  

2.24 Sun Jan 27 08:42:58 MST 2002
	- Playlists work again.  This was broken in version 2.22 when
		support for non-MP3 playlists added.

2.23 Thu Jan 24 18:44:42 EST 2002
	- Fixed problem with not being able to stream from virtual hosts.
	- Documented issues relating to not finding lame in the PATH when
		downsampling.

2.22 Sun Jan  6 15:25:07 EST 2002
	- Folded in Devin Carraway's <apache-modmp3@devin.com> patches
		for Ogg Vorbis Support.
	- Added Wave file support, along with a new architecture to deal
		with various media types.
	- Added Allen Day's modifications for handling directory list icons
		with some of my own modifications to make this more flexible.
	- Bug fixes in the ICY headers.

2.21 Tue Nov 20 11:44:43 EST 2001
	- Folded in Allen Day's modifications for
		handling .pls (audio/x-scpls) files.
	- Folden in Iaian Wade's <iwade@habitue.net> modifications for 
		handling ETags (speeds performance on big directories).
	- Changes from Joshua Cantard to fix information sent prior to streaming.

2.20 Mon Sun Sep 23 01:11:37 EDT 2001
	- Multiple bugs fixed in playlist.

2.19 Mon Jul 16 21:36:15 EDT 2001
	- Patch for Ask Bjoern Hansen to support the scroll bar when
		streaming to WinAmp.

2.18 Sun Jun 10 17:21:36 EDT 2001
	- Patch from Chris Nandor to get the file size approximately
		right when streaming downsampled files.
	- Patch from Chris Nandor to fix problems downloading
		MP3 files when resampling turned on.
	- Fixed an HTML-generating bug that prevented the browser window
		title from being set.
	- Removed hard-coded white background color from the resampling
		window.

2.17 Mon Feb 12 10:27:38 EST 2001
	- Fixed problem of multiple fetch buttons appearing when
		AllowStream is set to "no"

2.16 Mon Apr 30 22:31:17 EDT 2001
	- Added Apache::MP3::Resample, which provides dynamic downsampling
		for people with slower connections.
	- Added MS Media player to list of acceptable streaming clients.

2.15 Mon Jan  1 22:20:24 EST 2001
	- Fixed security bug which allowed people to bypass the AllowDownload setting.
	- Made "AllowDownload no" compatible with cover art.

2.14 Sat Dec 30 23:10:32 EST 2000
	- Fixed documentation to indicate that AddType requires dots before the
		.mp3 and .MP3 extension names.
	- Made the cache writing taint safe.
	- Fixed generation of extended playlists to work with earlier versions
		of WinAmp.
	- Added playlist and local streaming patches kindly provided by
		Jamie O'Shaughnessy.
	- Fixed display of directory path links when the current directory has same
		name as a parent directory.

2.13 Thu Nov 30 12:17:31 EST 2000
	- Patch from Jeffrey Sumler to prevent faults when streaming files containing
		regular expression metacharacters
	- Fix to prevent entire directory from being streamed when the directory
		has same name as the desired MP3 file (Jeffrey Sumler identified problem).

2.12 Tue Nov 21 08:39:15 EST 2000
	** IMPORTANT NOTE **
	I have normalized field names and added some new fields.  In particular, "kbps" has
	been renamed "bitrate".  If you are using the mp3info caching feature, please delete 
	your cache directory so that the new fields are correctly cached

	-Check for invalid songs in playlist and refuse to play them.
	-Added Jeffrey Seifert's patch for cover art.  You can now place a file named
		"cover.jpg" in an album subdirectory and have it appear next to
		the album's name.
	-Added Chris Nandor's patch for Accept-Range so that certain players can
		fetch selected portions of the song (in theory at least).
	-Added Kee Hinkley's extended playlist patch, which allows WinAmp (and other players?)
		to display the title and duration of the song in the playlist.
	-Added Gary Johnston's patch which fixes truncated songs when streaming from a
		Windows server (the old binmode() trap).

2.11 Sat Sep  9 18:04:55 EDT 2000
	Added Tim Ayer's patches for sorting.  Now can sort playlists 
		on multiple fields (nice!)
	Detect when playlist cookie is full, alert user, and suppress 
		further attempts to add to playlist.
	Documentation fixes.

2.10 Sun Sep  3 20:29:41 MET DST 2000
	Ooops.  Playlist.pm was inadvertently left out; now it's back in.
	Fixed divide by zero error on corrupt MP3 files.


Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME		=> 'Apache::MP3',
    AUTHOR              => 'Lincoln Stein <lstein@cshl.org>',
    ABSTRACT            => 'mod_perl class for generating browsable MP3 directory lists',
    ABSTRACT_FROM       => 'MP3.pm',
    VERSION             => '3.99',
    PREREQ_PM		=> {
			    MP3::Info             => 0.8,
			    Apache2::RequestRec   => 2.000001,
			    Audio::Wav            => 0.02,
			    Ogg::Vorbis::Header   => 0.02, # b/c Ogg::Vorbis can't be found by CPAN shell
			    Inline::MakeMaker     => 0.01, # required by Ogg::Vorbis, but not specified in it's makefile, causing CPAN shell install to fail.
			    Locale::Maketext      => 1.03,
			    CGI::Session          => 4.03,
			   }, # e.g., Module::Name => 1.1
);

1;


MP3/Sorted.pm
package Apache::MP3::Sorted;
# $Id$
# example of how to subclass Apache::MP3 in order to provide
# control over the sorting of the rows of the MP3 table

use strict;
use Apache::MP3;
use CGI qw/:standard *TR param/;

use vars qw(@ISA $VERSION);
@ISA = 'Apache::MP3';

$VERSION = 2.02;

use constant DEBUG => 0;

# to choose the right type of sort for each of the mp3 fields
my %sort_modes = (
		  # config        field       sort type
		  title       => [qw(title        alpha)],
		  artist      => [qw(artist       alpha)],
		  duration    => [qw(seconds      numeric)],
		  seconds     => [qw(seconds      numeric)],
		  genre       => [qw(genre        alpha)],
		  album       => [qw(album        alpha)],
		  comment     => [qw(comment      alpha)],
		  description => [qw(description  alpha)],
		  bitrate     => [qw(bitrate      numeric)],
		  filename    => [qw(filename     alpha)],
		  kbps        => [qw(kbps         numeric)],
		  samplerate  => [qw(samplerate   numeric)],
		  track       => [qw(track        numeric)],
		  year        => [qw(year         numeric)],
		 );

sub sort_fields {
  my $self       = shift;
  my $field_spec = lc param('sort') if param('sort');
  # If there is not SortFields var, look for SortField, which is the var used
  # by the one sort field Apache::MP3::Sorted. This should retain a good amount
  # of backward compatibility
  $field_spec  ||= lc($self->r->dir_config('SortFields')
                      || $self->r->dir_config('SortField'));
  return split /[^\w+-]+/, $field_spec;
}

# sort MP3s
sub sort_mp3s {
  my $self   = shift;
  my $files  = shift;
  my @sort_info = (); # holds info about how to sort each field

  # Verify the sorting fields and build up our @sort_info data structure that
  # keep track of how everything should work.
  foreach ($self->sort_fields) {
    my ($reverse_sort, $field) = /^([+-]?)(.*)/;
    $reverse_sort = ($reverse_sort eq '-');
    if (exists($sort_modes{$field})) {
      # each @sort_info entry is [sort_order, sort_field, sort_type]
      push(@sort_info, [$reverse_sort, $sort_modes{$field}[0], $sort_modes{$field}[1]]);
    } else {
      $self->r->warn("Ignoring unsupported sort field $_ in sort_mp3s().");
    }
  }

  if (!@sort_info) {
    # no known sort types given
    $self->r->warn("No recognized sort fields passed to sort_mp3s()") if DEBUG;
    return $self->SUPER::sort_mp3s($files);
  }

  # Create an anonymous subroutine so we have closure of @sort_info and $files
  my $by_fields_sorter =
    sub {
      my $val = 0;
      foreach (@sort_info) {
        my ($reverse_sort, $field, $sort_type) = @$_;
        my ($x, $y) = ($files->{$a}{$field}, $files->{$b}{$field});
        ($x, $y) = ($y, $x) if $reverse_sort; # swap args for reverse sort
        if ($sort_type eq 'alpha') {
          $val = $x cmp $y;
        } elsif ($sort_type eq 'numeric') {
          $val = $x <=> $y;
        }
        last if $val;
      }
      return $val; # Found no compare difference between elements
    };

  return sort $by_fields_sorter keys %$files;
}

sub format_table_fields {
  my $self = shift;
  my $sort = param('sort') || '';
  my @fields;

  foreach ($self->fields) {
    $_ ||= '';
    my $sort = $sort eq lc($_)  ? lc("-$_") : lc($_);
    push @fields,a({-href=>"?sort=$sort"},
      $self->x(ucfirst($_))
    );
  }
  @fields;
}

# Add hidden field for sorting
sub mp3_list_bottom {
  my $self = shift;
  print hidden('sort');
  $self->SUPER::mp3_list_bottom(@_);
}


1;

=head1 NAME

Apache::MP3::Sorted - Generate sorted streamable directories of MP3 files

=head1 SYNOPSIS

 # httpd.conf or srm.conf
 AddType audio/mpeg    mp3 MP3

 # httpd.conf or access.conf
 <Location /songs>
   SetHandler perl-script
   PerlHandler Apache::MP3::Sorted
   PerlSetVar  SortFields    Album,Title,-Duration
   PerlSetVar  Fields        Title,Artist,Album,Duration
 </Location>

=head1 DESCRIPTION

Apache::MP3::Sorted subclasses Apache::MP3 to allow for sorting of MP3
listings by various criteria.  See L<Apache::MP3> for details on
installing and using.

=head1 CUSTOMIZING

This class adds one new Apache configuration variable, B<SortFields>.
This is a list of the names of the fields to sort by default when the
MP3 file listing is first displayed, after which the user can change
the sort field by clicking on the column headers.

=over 4

=item SortFields I<field>

The value of B<SortFields> may contain the names of any of the fields
in the listing, such as I<Title>, or I<Album> I<Duration>.  By
default, the sort direction will be alphabetically or numerically
ascending.  Reverse this by placing a "-" in front of the field name
(for symmetry, you can also prepend a "+" to signify normal ascending
order).  After the initial sort, the user can change the sort order by
clicking on the column headings of the file listing table.

Examples:

  PerlSetVar SortFields  Album,Title    # sort ascending by album, then title
  PerlSetVar SortFields  +Artist,-Kbps  # sort ascending by artist, descending by kbps

When constructing a playlist from a recursive directory listing,
sorting will be B<global> across all directories.  If no sort order is
specified, then the module reverts to sorting by file and directory
name.  A good value for SortFields is to sort by Artist,Album and
track:

  PerlSetVar SortFields Artist,Album,Track

Alternatively, you might want to sort by Description, which
effectively sorts by title, artist and album.

The following are valid fields:

    Field        Description

    album	 The album
    artist       The artist
    bitrate      Streaming rate of song in kbps
    comment      The comment field
    description	 Description, as controlled by DescriptionFormat
    duration     Duration of the song in hour, minute, second format
    filename	 The physical name of the .mp3 file
    genre        The genre
    samplerate   Sample rate, in KHz
    seconds      Duration of the song in seconds
    title        The title of the song
    track	 The track number

Field names are case insensitive.

=back

=head1 METHODS

Apache::MP3::Sorted overrides the following methods:

 sort_mp3s()  mp3_table_header()   mp3_list_bottom()

It adds one new method:

=over 4

=item $field = $mp3->sort_fields

Returns a list of the names of the fields to sort on by default.

=back

=head1 BUGS

Let me know.

=head1 AUTHOR

Copyright 2000, Lincoln Stein <lstein@cshl.org>.

This module is distributed under the same terms as Perl itself.  Feel
free to use, modify and redistribute it as long as you retain the
correct attribution.

=head1 ACKNOWLEDGEMENTS

Tim Ayers <tayers@bridge.com> conceived and implemented the multiple
field sorting system.

=head1 SEE ALSO

L<Apache::MP3>, L<MP3::Info>, L<Apache>


=cut


MP3/Resample.pm
package Apache::MP3::Resample;

# $Id$
# Resamples (downsamples) on the fly

use strict;
use vars qw(@ISA $VERSION);
use Apache2::Const -compile => qw(:common);
use IO::File;
use CGI qw(:standard *table *TR *td escape);
use CGI::Cookie;
use Apache::MP3::Playlist;
use File::Basename;
$VERSION = 1.0;

@ISA = 'Apache::MP3::Playlist';

use constant ENCODE => qq(%Dlame%I %b -%F);

my @DECOMPRESSOR_DEFAULTS = ("audio/shorten" => "shorten -x %f -|",
			     "audio/flac" => "flac -d -c -s %f|");

my @PRESET_DEFAULTS = (
	       '24 kbps' => '-b24 --lowpass 4.0  --highpass 0.1  --resample 8',
	       '40 kbps' => '-b40 --lowpass 4.0  --resample 16',
	       '56 kbps' => '-b56 --lowpass 12.0 --resample 22.05',
	       '64 kbps' => '-b64 --lowpass 15.0 --resample 22.05',
	       '96 kbps' => '-b96 --lowpass 15.0',
	       '128 kbps' => '-b128 --lowpass 18',
	       '160 kbps' => '-b160',
	       '192 kbps' => '-b192',
	      );

my (%PRESETS,@PRESETS,%DECOMPRESSORS,@DECOMPRESSORS);

sub new {
   my $proto = shift;
   my $class = ref ($proto) || $proto;
   my $self = $class->SUPER::new (@_);

   # Add to suffix list and supported type hash
   push (@{$self->{'suffixes'}}, ".shn", ".SHN", ".flac", ".FLAC");
   $self->{'supported_types'}->{'audio/shorten'} = 'read_shorten';
   $self->{'supported_types'}->{'audio/flac'} = 'read_flac';

   bless ($self, $class);
   $self;
}

sub run {
  my $self = shift;
  $self->process_cookies;
  $self->SUPER::run();
}

sub process_cookies {
  my $self = shift;
  my $r = $self->r;

  # don't set cookies for ordinary MP3 file downloads.
  return if !param() && !-d $r->filename;

  if (my $cookies = CGI::Cookie->parse($r->headers_in('Cookie'))) {
    $self->bitrate($cookies->{bitrate}->value)
      if $cookies->{bitrate};
  }
  $self->bitrate(param('bitrate')) if defined param('bitrate');
  my $c = CGI::Cookie->new(-name  => 'bitrate',
			   -value => $self->bitrate,
			   -expires => '+90d'
			  );
  tied(%{$r->err_headers_out})->add('Set-Cookie' => $c);
  param(bitrate => $self->bitrate);
}

sub bitrate {
  my $self = shift;
  my $g    = $self->{bitrate};
  $self->{bitrate} = shift if @_;
  return unless $self->presets($g);
  $g;
}

sub stream_parms {
  my $self = shift;
  my $p =  $self->SUPER::stream_parms;
  my $rate = escape($self->bitrate);
  $p .= ";bitrate=$rate" if $rate;
  $p;
}

sub presets {
  my $self = shift;
  unless (%PRESETS) {
    my @p;
    if (my $conf = $self->r->dir_config('ResamplePresets')) {
      @p = split /\s*(?:=>|,)\s*/,$conf;
    } else {
      @p = @PRESET_DEFAULTS;
    }
    my @i = map {$_*2} (0..(@p-1)/2);
    %PRESETS    = @p;
    @PRESETS = @p[@i];
  }
  my $bitrate = shift or return @PRESETS;
  return $PRESETS{$bitrate};
}

sub decompressor {
  my $self = shift;
  unless (%DECOMPRESSORS) {
    my @p;
    if (my $conf = $self->r->dir_config('DecompressorPresets')) {
      @p = split /\s*(?:=>|,)\s*/,$conf;
    } else {
      @p = @DECOMPRESSOR_DEFAULTS;
    }
    my @i = map {$_*2} (0..(@p-1)/2);
    %DECOMPRESSORS    = @p;
    @DECOMPRESSORS = @p[@i];
  }
  my $mime = shift or return @DECOMPRESSORS;
  return $DECOMPRESSORS{$mime};
}

sub sample_popup {
  my $self = shift;
  my @bitrates = $self->presets;
  unshift @bitrates,0;
  my %labels = (0 => '--');
  return (
	  start_form(-name=>'form'),
	  table(
		TR(th('Resample:'),
		   td(
		      popup_menu(-name  => 'bitrate',
				 -value => \@bitrates,
				 -labels => \%labels,
				 -onChange => 'form.submit()'
				),
		     )
		  )
		),
	    end_form);
}

sub directory_top {
  my $self = shift;
  print start_table({-width=>'100%'});
  print start_TR,start_td;
  $self->SUPER::directory_top(@_);
  print end_td;
  print td({-align=>'RIGHT',-valign=>'TOP'},
	   $self->sample_popup());
  print end_TR,end_table;
}

sub open_file {
  my $self = shift;
  return $self->SUPER::open_file(@_) unless $self->bitrate;
  my $file = shift;
  my $bitrate = $self->bitrate;
  my $presets = $self->presets($bitrate);
  my $type = $self->r->lookup_file($file)->content_type;
  my $decompress = $self->decompressor ($type) || "";
  my $encode = $self->r->dir_config('MP3Encoder') || ENCODE;
  my $inputtype = length $decompress ? "" : " --mp3input";
  my $percentF = length $decompress ? "" : ("<" . quotemeta ($file));
  $decompress =~ s{%([a-zA-Z])}
    { $1 eq 'f' ? quotemeta($file) :
	"%$1"}exg;
  $encode =~ s{%([a-zA-Z])}
              {$1 eq 'b' ? $presets         :
	       $1 eq 'D' ? $decompress      :
	       $1 eq 'I' ? $inputtype       :
	       $1 eq 'F' ? $percentF        :
               $1 eq 'f' ? quotemeta($file) :
               "%$1"}exg;
  my $filter = $self->r->dir_config('VerboseMP3Encoder')
    ? "$encode |" : "$encode 2>/dev/null |";
  return IO::File->new($filter);
}

sub read_shorten {
  my $self = shift;
  my ($file,$data) = @_;
  my $sec;
  my $bits = 0;
  my $samples = 0;
  my $channels = 0;
  my $track = "";

  open (SHNINFO, "shntool info \Q$file\E 2>/dev/null|") or return;

  while (<SHNINFO>) {
     if (/^length:\s+(\d+):(\d+)\.(\d+)/) {
	$sec = $1 * 60 + $2;
     } elsif (/^channels:\s+(\d+)/) {
	$channels = $1;
     } elsif (/^samples\/sec:\s+(\d+)/) {
	$samples = $1;
     } elsif (/^bits\/sample:\s+(\d+)/) {
	$bits = $1;
     }
  }
  close SHNINFO;

  %$data = (
	    title  => basename ($file),
	    artist => "",
	    album  => "",
	    year   => "",
	    genre  => "",
	    track  => $track,
	    comment => "",
	    min         => int $sec/60,
	    sec         => $sec % 60,
	    seconds     => $sec,
	    bitrate     => int ($channels * $bits * $samples / 1024),
	    samplerate  => $samples,
	    duration    => sprintf("%dm %2.2ds", int $sec/60,$sec%60),
	   )
}

sub read_flac {
  my $self = shift;
  my ($file,$data) = @_;
  my $sec;
  my $bits = 0;
  my $samples = 0;
  my $channels = 0;
  my $track = "";
  my $rate = 0;

  open (METAFLAC, "metaflac --list --block-number 0 \Q$file\E 2>/dev/null|")
    or return;

  while (<METAFLAC>) {
     if (/sample_rate: (\d+)\s/) {
	$rate = $1;
     } elsif (/channels: (\d+)/) {
	$channels = $1;
     } elsif (/bits-per-sample: (\d+)/) {
	$bits = $1;
     } elsif (/total samples: (\d+)/) {
	$samples = $1;
	$sec = $samples / $rate if $rate;
     }
  }
  close METAFLAC;

  %$data = (
	    title  => basename ($file),
	    artist => "",
	    album  => "",
	    year   => "",
	    genre  => "",
	    track  => $track,
	    comment => "",
	    min         => int $sec/60,
	    sec         => $sec % 60,
	    seconds     => $sec,
	    bitrate     => int ($channels * $bits * $rate / 1024),
	    samplerate  => $samples,
	    duration    => sprintf("%dm %2.2ds", int $sec/60,$sec%60),
	   )
}

1;

=head1 NAME

Apache::MP3::Resample - Downsample MP3/FLAC/Shorten files during streaming

=head1 SYNOPSIS

 # httpd.conf or access.conf
 AddType audio/shorten .shn .SHN
 AddType audio/flac .flac

 Alias /apache_mp3 /usr/share/libapache-mp3-perl

 <Location /songs>
   SetHandler perl-script
   PerlHandler Apache::MP3::Resample
   PerlSetVar CacheDir		/var/cache/Apache::MP3
   PerlSetVar AllowDownload	no
   PerlSetVar SortFields	Album,Track,Title,-Duration
   PerlSetVar Fields		Track,Title,Artist,Album,Duration,Bitrate
 </Location>

=head1 DESCRIPTION

Apache::MP3::Resample subclasses Apache::MP3::Playlist to allow the
user to downsample audio files before streaming them.  This allows users
on slower connections to stream songs.  When this module is installed,
a menu of bitrates is presented in the upper right-hand corner of the
screen.  The user can choose from one of the bitrates, or select a
mode that performs no resampling.  The selected bitrate is maintained
in a persistent cookie so that resampling is performed whenever the
user returns to the site.

This module requires a command-line MP3 encoder to resample and
reencode the audio data.  If not otherwise specified,
Apache::MP3::Resample will try to use the Open Source Lame MP3 encoder.
This utility is available at http://www.sulaco.org/mp3.  Version 3.90
was used during the development of this module.  Your results with
other versions may vary.

When you install Lame (or the encoder of your choice), be sure to
place it in a directory located in Apache's PATH so that the module
can find them at run time.  You may need to set the PATH environment
variable during Apache's launch, or by explicitly adding a B<SetEnv>
directive to the Apache configuration file.

You should be aware that the decoding/reencoding process is
CPU-intensive, and server performance may degrade as the number of
simultaneous users increases.

=head1 CUSTOMIZATION

This class recognizes the following two configuration variables in
addition to those recognized by its superclasses.

=over 4

=item MP3Encoder

The command to use to invoke the MP3 encoder.  It should accept CDDA
data on standard input and write MP3 data to standard output.  The
command should contain the replacement sequences b<%b> and b<%f>.  At
run time, b<%b> will be replaced with the options used to set the
bitrate and sampling frequency, while b<%f> will contain the MP3 file
to be streamed.

If not present, the following default is used:

  lame --mp3input %b - <%f

=item DecompressorPresets

A list of MIME types and decompression programs to use for each.  The
format uses a variant of the standard Perl hash format, in which the
keys are the MIME types for the audio files and the values are the
command which will be used to decompress files of that type.  The
token "%f" will be replaced by the name of the compressed audio file
and the command should end with a pipe (|) character.  Here is a
simple example, which is also the default list:

    PerlSetVar  DecompressorPresets  '"audio/shorten" => "shorten -x %f -|",\
                                      "audio/flac" => "flac -d -c -s %f|"'

Note the use of quotation marks and backslashes to protect whitespace
and newlines respectively.

=item ResamplePresets

A list of bitrates and the command-line options to pass to the
encoder.  The format uses a variant of the standard Perl hash format,
in which the keys are the bitrates to present to the user and the
values are command-line options to pass to the encoder.  Here is a
simple example:

    PerlSetVar  ResamplePresets  '16 kbps => -b16,\
	                          56 kbps => -b56,\
                                 128 kbps => -b128,\
				 160 kbps => -b160'

Note the use of quotation marks and backslashes to protect whitespace
and newlines respectively.

Here is another example, which takes advantage of the --preset feature
present in newer versions of Lame.

    PerlSetVar  ResamplePresets 'phone => --preset phone,\
	                         voice  => --preset voice,\
                                 fm     => --preset fm, \
				 tape   => --preset tape,\
	                         hifi   => --preset hifi,\
                                 cd     => --preset cd'

The user will see a popup menu containing the entries "16 kbps", "56
kbps" and so forth, as well as a blank ("--") entry that is provided
automatically.  Upon selecting each option the corresponding
command-line arguments will be slotted into the "%s" variable in the
encoder line specified by B<MP3Encoder>.

More complex command line options are possible.  For example, to
invoke VBR (variable bitrate) encoding and resample the output to
22.05 kHz, you could apply these options

    PerlSetVar  ResamplePresets  '16 => -b16 -v --resample 22.05,\
	                          56 => -b56 -v --resample 22.05,\
                                 128 => -b128 -v --resample 22.05,\
				 160 => -b160 -v --resample 22.05'

Lame only accepts certain combinations of command-line options, and I
do not fully understand the restrictions.  Please do not e-mail me
with Lame-related questions.

The default presets are:

     24  kbps => -b24 --lowpass 4.0  --highpass 0.1  --resample 8,
     40  kbps => -b40 --lowpass 4.0  --resample 16,
     56  kbps => -b56 --lowpass 12.0 --resample 22.05,
     64  kbps => -b64 --lowpass 15.0 --resample 22.05,
     96  kbps => -b96 --lowpass 15.0,
     128 kbps => -b128 --lowpass 18,
     160 kbps => -b160,
     192 kbps => -b192

=back

=head1 METHODS

This module overrides the inherited run(), open_file() and
directory_top() methods.  It adds the following new methods:

=over 4

=item bitrate()

Set or get the command-line options to pass to the encoder for a
desired bitrate.

=item stream_parms()

Return the parameters to append to an MP3 playlist entry in order to
activate resampling.

=item presets()

Get the names of the bitrate presets specified by the ResamplePresets
configuration variable, or the default names.  Called with an argument
equal to the name of a preset, returns the command-line arguments to
pass to the encoder.

=item sample_popup()

Draws the popup menu with the sample rate options.

=back

=head1 BUGS

When the external program is invoked to downsample the MP3 data, its
standard error is redirected to /dev/null.  This prevents Lame's
informational messages from gumming up the server error log, but also
prevents the system from giving you helpful diagnostic messages, such
as "file not found".  If you are having trouble with the downsampling,
set the configuration variable VerboseMP3Encoder to a true value in
order to see the standard error messages.

Also, the module does not function properly unless CacheDir is set and
points to a directory which exists and is writeable by the Apache
server.

=head1 ACKNOWLEDGEMENTS

Many people have requested this feature and/or proposed implementations.
Thank you all for your help.

FLAC and Shorten support added by Caleb Epstein <cae@bklyn.org>.

=head1 AUTHOR

Copyright 2000, Lincoln Stein <lstein@cshl.org>.

This module is distributed under the same terms as Perl itself.  Feel
free to use, modify and redistribute it as long as you retain the
correct attribution.

=head1 SEE ALSO

L<Apache::MP3::Playlist>, L<Apache::MP3>, L<MP3::Info>, L<Apache>

=cut

